### DISCUSSION TITLE
How to use QRImageWithLogo.php

### DISCUSSION THREAD

**Discussion by Flaschenzug** [chillerlan/php-qrcode#220] [Q&A]
I am having issues with using the QRImageWithLogo.php ( https://github.com/chillerlan/php-qrcode/blob/f012a15bff005c028ee1428b5831344d89879180/examples/QRImageWithLogo.php )

The basic QR Code generation is working.

`$data = 'https://URLTOPAGE';
echo '<img src="'.(new QRCode)->render($data).'" width="200" alt="QR Code" />';`

But when I try to use the WithLogo version (from the example above):

`use chillerlan\QRCode\QRCode;
use chillerlan\QRCode\Output\{QRCodeOutputException, QRImage};
require_once '/vendor/autoload.php';

class QRImageWithLogo extends QRImage{   
   public function dump(string $file = null, string $logo = null):string{
   ... 
}

echo '<img src="'.(new QRImageWithLogo(null, null))->render($data).'" width="200" alt="QR Code" />';`

Fatal error: Uncaught Error: chillerlan\QRCode\Output\QRGdImage::__construct(): Argument #1 ($options) must be of type chillerlan\Settings\SettingsContainerInterface

And I have no idea, how I give QRImageWithLogo() what it needs.
I tried to find information about it here: https://php-qrcode.readthedocs.io/en/main/ 
But I am stuck.

Thanks for your help

**Comment by codemasher** [comment]
Hi! It löoks like you're using a v4.x example with the dev-main or beta branch (the class `QRGdImage` in which the error is thrown doesn't exist in v 4.x) - the documentation on readthedocs is currently only for the v5 (main) branch.

It seems the output instance is not invoked properly - it should look like this, where you replace `MyCustomOutput::class` with the class name of your custom class:

https://github.com/chillerlan/php-qrcode/blob/f012a15bff005c028ee1428b5831344d89879180/examples/custom_output.php#L31-L36

**Reply by codemasher** [reply]
> Fatal error: Uncaught Error: chillerlan\QRCode\Output\QRGdImage::__construct(): Argument 1 ($options) must be of type chillerlan\Settings\SettingsContainerInterface

Also, the error might be thrown because the newer php-qrcode version uses a different version of the `SettingsContainerInterface` class if running under PHP 8.1+.

**Comment by Flaschenzug** [comment]
Okay, thanks. Using your first way (new QROptions with the MyCustomOutput) I get a qr code.
The only problem: The logo is not displayed inside the QR Code. QR code is correct (scannable).

As a path I used: ABSPATH."path-to-image.png";
Any ideas how to solve this?

And yes: I use PHP 8.1
So should I upgrade to V5 ?

And one last question:
I tried to add 'logo' to the options class list, but it does not work. Same for the widht and height of the logo.
How do I achive that using the method you showed above ($options = new QROptions([ ... )

Thank you so much :-)

**Reply by codemasher** [reply]
> As a path I used: ABSPATH."path-to-image.png"; Any ideas how to solve this?

From what i gathered `ABSPATH` is a Wordpress predefined constant "that represents the absolute path to the root directory of your WordPress installation". So you need to make sure your image is somewhere in there.

> So should I upgrade to V5 ?

Yes, absolutely. At this point you can even use `dev-main` which differs slightly from the beta-release. I'm currently busy to finalize the v5 release and hope to get it out in the next few weeks, hopefully around PHP 8.3.

> I tried to add 'logo' to the options class list, but it does not work. Same for the width and height of the logo. How do I achive that using the method you showed above

For that you might want to check out [the Imagick example](https://github.com/chillerlan/php-qrcode/blob/main/examples/imagickWithLogo.php), which extends the options class to achieve this. It also contains a setter method that checks for the existence of the given image file, which solves your problem mentioned above.

**Reply by Flaschenzug** [reply]
The path to the image is correct (and yes, it is wordpress).
And if I check for the width and height of the loaded logo, I get the right dimensions.
So the logo is loaded correctly.

But not displayed inside the white square of the QR Code :-/

**Reply by codemasher** [reply]
It would help if you post the full code of your custom classes - remote diagnosis is *hard*-

**Reply by Flaschenzug** [reply]
```
use chillerlan\QRCode\QRCode;
use chillerlan\QRCode\QROptions;
use chillerlan\QRCode\Output\{QRCodeOutputException, QRImage};
require_once get_stylesheet_directory() . "/assets/external/vendor/autoload.php";

class QRImageWithLogo extends QRImage{

	public function dump(string $file = null, string $logo = null):string{
		// set returnResource to true to skip further processing for now
		$this->options->returnResource = true;


        $logo = ABSPATH."correct-path-to-file.png";

		// of course you could accept other formats too (such as resource or Imagick)
		// i'm not checking for the file type either for simplicity reasons (assuming PNG)
		if(!is_file($logo) || !is_readable($logo)){
			throw new QRCodeOutputException('invalid logo');
		}

		$this->matrix->setLogoSpace(
			//$this->options->logoWidth,
			//$this->options->logoHeight
                        15,15
		// not utilizing the position here
		);

		// there's no need to save the result of dump() into $this->image here
		parent::dump($file);

		$im = imagecreatefrompng($logo);

		// get logo image size
		$w = imagesx($im);
		$h = imagesy($im);

		// set new logo size, leave a border of 1 module
		$lw = ($this->options->logoWidth - 2) * $this->options->scale;
		$lh = ($this->options->logoHeight - 2) * $this->options->scale;

		// get the qrcode size
		$ql = $this->matrix->size() * $this->options->scale;

		// scale the logo and copy it over. done!
		imagecopyresampled($this->image, $im, ($ql - $lw) / 2, ($ql - $lh) / 2, 0, 0, $lw, $lh, $w, $h);

		$imageData = $this->dumpImage();

		if($file !== null){
			$this->saveToFile($imageData, $file);
		}

		if($this->options->imageBase64){
			$imageData = 'data:image/'.$this->options->outputType.';base64,'.base64_encode($imageData);
		}

		return $imageData;
	}

}

$options = new QROptions([
	'eccLevel'        => QRCode::ECC_H,
	'outputType'      => QRCode::OUTPUT_CUSTOM,
	'outputInterface' => QRImageWithLogo::class,
]);
$data = 'https://test.com';
echo '<img src="'.(new QRCode($options))->render($data).'" width="300" alt="QR Code" />';```

**Reply by codemasher** [reply]
Hmm, so far it looks ok, the `$this->options->logoWidth` in the width and height calculation however will return `null` as this property doesn't exist - in the original code it is derived from the logo *space* width there. If you replace these with the proper variables `$this->options->logoSpaceWidth` and `$this->options->logoSpaceHeight` and set these in the `QROptions` call below, the logo appears.

**Reply by codemasher** [reply]
Also, there's an error on my end in the examples where the image is encoded into base64: `$this->options->outputType` there will insert "custom" where it should be PNG, however, in the original example i circumvented this by manually invoking the output instance and set the output type to `QRCode::IMAGE_PNG`.

**Reply by Flaschenzug** [reply]
I changed it .. now i get an error:

> Fatal error: Uncaught Error: Undefined constant chillerlan\QRCode\QRCode::IMAGE_PNG

```
use chillerlan\QRCode\QRCode;
use chillerlan\QRCode\QROptions;
use chillerlan\QRCode\Output\{QRCodeOutputException, QRImage};
require_once get_stylesheet_directory() . '/assets/external/vendor/autoload.php';

class QRImageWithLogo extends QRImage{

	/**
	 * @param string|null $file
	 * @param string|null $logo
	 *
	 * @return string
	 * @throws \chillerlan\QRCode\Output\QRCodeOutputException
	 */
	public function dump(string $file = null, string $logo = null):string{
		// set returnResource to true to skip further processing for now
		$this->options->returnResource = true;
		$this->options->logoSpaceWidth = 15;
		$this->options->logoSpaceHeight = 15;
		$this->options->logoWidth = 15;
		$this->options->logoHeight = 15;
        $logo = ABSPATH."logo.png";

		// of course you could accept other formats too (such as resource or Imagick)
		// i'm not checking for the file type either for simplicity reasons (assuming PNG)
		if(!is_file($logo) || !is_readable($logo)){
			throw new QRCodeOutputException('invalid logo');
		}

		$this->matrix->setLogoSpace(
			//$this->options->logoWidth,
			//$this->options->logoHeight
            15,
            15
		// not utilizing the position here
		);

		// there's no need to save the result of dump() into $this->image here
		parent::dump($file);

		$im = imagecreatefrompng($logo);

		// get logo image size
		$w = imagesx($im);
		$h = imagesy($im);

		// set new logo size, leave a border of 1 module
		$lw = ($this->options->logoWidth - 2) * $this->options->scale;
		$lh = ($this->options->logoHeight - 2) * $this->options->scale;

		// get the qrcode size
		$ql = $this->matrix->size() * $this->options->scale;

		// scale the logo and copy it over. done!
		imagecopyresampled($this->image, $im, ($ql - $lw) / 2, ($ql - $lh) / 2, 0, 0, $lw, $lh, $w, $h);

		$imageData = $this->dumpImage();

		if($file !== null){
			$this->saveToFile($imageData, $file);
		}

		if($this->options->imageBase64){
			$imageData = 'data:image/'.$this->options->outputType.';base64,'.base64_encode($imageData);
		}

		return $imageData;
	}

}

$options = new QROptions([
	//'version'         => 4,
	'eccLevel'        => QRCode::ECC_H,
	'outputType'      => QRCode::IMAGE_PNG,
	'outputInterface' => QRImageWithLogo::class,
]);


$data = 'https://test.com';
echo '<img src="'.(new QRCode($options))->render($data).'" width="300" alt="QR Code" />';
```

**Reply by codemasher** [reply]
Ah my bad i forgot what the name was `OUTPUT_IMAGE_PNG` - but it is deprecated anyway as of v5 and you should be using `QROutputInterface::GDIMAGE_PNG`. Try staring from the example that's shipped with the version you installed - mix-and match isn't the best idea here as there were quite a few internal changes betweeen the major versions.

**Reply by Flaschenzug** [reply]
Still no Icon inside and not even a white space anymore :-).
So as recommended I will try to use V5 dev-main. I'll send an update.

**Reply by Flaschenzug** [reply]
I upgraded to dev-main (using composer) and tried it again, using the instructions from the documentation and your comments here. But the problem stays the same. I get a QR code (no white square in the middle, no logo):

```
use chillerlan\QRCode\QRCode;
use chillerlan\QRCode\QROptions;
use chillerlan\QRCode\Output\{QRCodeOutputException, QRImage};
require_once get_stylesheet_directory() . '/assets/external/vendor-new/autoload.php';

use function imagecopyresampled, imagecreatefrompng, imagesx, imagesy, is_file, is_readable;

/**
 * @property \chillerlan\QRCodeExamples\LogoOptions $options
 */
class QRImageWithLogo extends QRImage{

	/**
	 * @param string|null $file
	 * @param string|null $logo
	 *
	 * @return string
	 * @throws \chillerlan\QRCode\Output\QRCodeOutputException
	 */
	public function dump(string $file = null, string $logo = null):string{
		// set returnResource to true to skip further processing for now
		$this->options->returnResource = true;
		$this->options->logoSpaceWidth = 15;
		$this->options->logoSpaceHeight = 15;
		$this->options->logoWidth = 15;
		$this->options->logoHeight = 15;
		$logo = ABSPATH."logo.png";

		// of course you could accept other formats too (such as resource or Imagick)
		// i'm not checking for the file type either for simplicity reasons (assuming PNG)
		if(!is_file($logo) || !is_readable($logo)){
			throw new QRCodeOutputException('invalid logo');
		}

		$this->matrix->setLogoSpace(
			15,
            15
		);

		// there's no need to save the result of dump() into $this->image here
		parent::dump($file);

		$im = imagecreatefrompng($logo);

		// get logo image size
		$w = imagesx($im);
		$h = imagesy($im);

		// set new logo size, leave a border of 1 module
		$lw = ($this->options->logoWidth - 2) * $this->options->scale;
		$lh = ($this->options->logoHeight - 2) * $this->options->scale;

		// get the qrcode size
		$ql = $this->matrix->size() * $this->options->scale;

		// scale the logo and copy it over. done!
		imagecopyresampled($this->image, $im, ($ql - $lw) / 2, ($ql - $lh) / 2, 0, 0, $lw, $lh, $w, $h);

		$imageData = $this->dumpImage();

		if($file !== null){
			$this->saveToFile($imageData, $file);
		}

		if($this->options->imageBase64){
			$imageData = 'data:image/'.$this->options->outputType.';base64,'.base64_encode($imageData);
		}

		return $imageData;
	}

}

$options = new QROptions([
	'version'         => 5,
	'eccLevel'        => QRCode::ECC_H,
	'outputType'      => QRCode::OUTPUT_IMAGE_PNG,
	'outputInterface' => QRImageWithLogo::class,
]);

$data = 'https://google.com/';
echo '<img src="'.(new QRCode($options))->render($data).'" width="300" alt="QR Code" />';
```

If I change it to the original code using OUTPUT_CUSTOM, I get a QR Code with a blank square in the middle (also no image loaded):

```
$options = new QROptions([
	'version'         => 5,
	'eccLevel'        => QRCode::ECC_H,
	'outputType'      => QRCode::OUTPUT_CUSTOM,
	'outputInterface' => QRImageWithLogo::class,
]);
```

Am i missing something? Thanks


composer.json file
```
{
    "require": {
	"php": "^7.4 || ^8.0",
        "smhg/sepa-qr-data": "^1.2",
        "chillerlan/php-qrcode": "dev-main"
    }
}
```

**Reply by codemasher** [reply]
Again, you are setting the non-existent options "logoWidth" and "logoHeight" - these have zero effect as the property doesn't exist in the options class. In addition, as of PHP 8.2 you should be getting an error "Deprecated: Implicit conversion from float 117.5 to int loses precision" which is caused by the non-existent values.

Either change these variables to local or extend the QROptions class and add these properties.

**Reply by Flaschenzug** [reply]
Okay, finally I got it. 

I you should have the same problem:
Update to the newest version (5).

And don't use this demo: https://github.com/chillerlan/php-qrcode/blob/f012a15bff005c028ee1428b5831344d89879180/examples/QRImageWithLogo.php

Download the newest version: https://github.com/chillerlan/php-qrcode/releases/tag/5.0-beta
And look for the file /examples/imageWithLogo.php

Thanks for your incredibly fast help @codemasher !!!

**Reply by codemasher** [reply]
The issue has nothing to do with the used version per-se, it's rather a version mismatch (use the examples provided with the installed version). That said, you may want to fix several constants and the extended class from `QRImage` to `QRGdImage` as these are still remnants from the 4.x example and that are deprecated in v5.

**Reply by Flaschenzug** [reply]
And now I have another issue when using a long URL:
`Fatal error: Uncaught Exception: code length overflow. (660 > 368 bit)`

I found this topic: https://github.com/chillerlan/php-qrcode/issues/31
But I have no idea where to set QRCode::VERSION_AUTO

( I use this file: /examples/imageWithLogo.php )

```
$options = new QROptions([
	'version'             => 5,
	'eccLevel'            => EccLevel::H,
	'imageBase64'         => false,
	'addLogoSpace'        => true,
	'logoSpaceWidth'      => 13,
	'logoSpaceHeight'     => 13,
	'scale'               => 6,
	'imageTransparent'    => false,
	'drawCircularModules' => true,
	'circleRadius'        => 0.45,
	'keepAsSquare'        => [QRMatrix::M_FINDER, QRMatrix::M_FINDER_DOT],
]);

$qrcode = new QRCode($options);
$qrcode->addByteSegment('https://github.com/chillerlan/php-qrcode/blob/main/src/QROptionsTrait.php#L17-L24');
```

**Reply by codemasher** [reply]
For "version"? You can also simply omit the version setting, which triggers auto detection (it is the default value).

**Reply by codemasher** [reply]
I think most of this is written in [the documentation](https://php-qrcode.readthedocs.io/en/main/) - if you think the docs are still lacking information, feel free to propose a change.

**Reply by Flaschenzug** [reply]
Wonderfull. Removing the version works. Thank you!
